src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/PreferAsSpanOverSubstring.cs

```diff
@@ -0,0 +1,294 @@
+                    ++substringCalls;
+                }
+            }
+
+            if (substringCalls == 0)
+                return ImmutableArray<IMethodSymbol>.Empty;
+
+            //  Find all overloads that are tied for being the "best" overload. An overload is considered
+            //  "better" if it allows more Substring calls to be replaced with AsSpan calls.
+            var bestCandidates = ImmutableArray.CreateBuilder<IMethodSymbol>();
+            int resultQuality = 0;
+            var candidates = GetAllAccessibleOverloadsIncludingSelf(invocation, cancellationToken);
+            foreach (var candidate in candidates)
+            {
+                int quality = EvaluateCandidateQuality(symbols, isSubstringLookup, invocation, candidate);
+
+                //  Reject candidates that do not replace at least one Substring call.
+                if (quality < 1)
+                {
+                    continue;
+                }
+                else if (quality == resultQuality)
+                {
+                    bestCandidates.Add(candidate);
+                }
+                else if (quality > resultQuality)
+                {
+                    resultQuality = quality;
+                    bestCandidates.Clear();
+                    bestCandidates.Add(candidate);
+                }
+            }
+
+            return bestCandidates.ToImmutable();
+
+            //  Returns a number indicating how good the candidate method is. 
+            //  If the candidate is valid, the number of Substring calls that can be replaced with AsSpan calls is returned.
+            //  If the candidate is invalid, -1 is returned.
+            static int EvaluateCandidateQuality(in RequiredSymbols symbols, ReadOnlySpan<bool> isSubstringLookup, IInvocationOperation invocation, IMethodSymbol candidate)
+            {
+                var method = invocation.TargetMethod;
+
+                if (candidate.Parameters.Length != method.Parameters.Length)
+                    return -1;
+
+                int replacementCount = 0;
+                foreach (var parameter in candidate.Parameters)
+                {
+                    if (isSubstringLookup[parameter.Ordinal] && SymbolEqualityComparer.Default.Equals(parameter.Type, symbols.ReadOnlySpanOfCharType))
+                    {
+                        ++replacementCount;
+                        continue;
+                    }
+
+                    var oldParameter = method.Parameters[parameter.Ordinal];
+                    if (!SymbolEqualityComparer.Default.Equals(parameter.Type, oldParameter.Type))
+                    {
+                        return -1;
+                    }
+                }
+
+                return replacementCount;
+            }
+        }
+
+        private static IEnumerable<IMethodSymbol> GetAllAccessibleOverloadsIncludingSelf(IInvocationOperation invocation, CancellationToken cancellationToken)
```

Comment:
Can we use this extension method? https://github.com/dotnet/roslyn-analyzers/blob/1fa83fe751c6e245f20ab57cd863b5d17098520b/src/Utilities/Compiler/Extensions/IMethodSymbolExtensions.cs#L441

Reply:
That method is poorly named. Basically, it looks at a particular invocation operation and finds all overloads of the invoked method that would also be legal at that location, taking into account the type of the receiver (for instance methods) or type (for static methods). It's pretty specific to this analyzer; I can't make a version that uses an `IMethodSymbol` instead of an `IInvocationOperation` because it depends on the receiver of the invocation. I'll rename it to something more descriptive to make this clearer. 

If we did want to add an extension method that does this, it couldn't go in `IMethodSymbolExtensions` because the extended receiver would need to be `IInvocationOperation`.